### PR TITLE
refactor: `mimirtool analyze prometheus`: add concurrency and resiliency

### LINKED ISSUE #3062: mimir-tool: analyse prometheus: ux: refactor / resiliency / concurrency
[analyse_prometheus.go](https://github.com/grafana/mimir/blob/main/pkg/mimirtool/commands/analyse_prometheus.go) is not cleanly written and contains lots of duplicate logics. Also it does not have any resiliency and concurrency mechanism if we run against very large Prometheus instance. (it takes a few hours or longer). Here are my two cents:

* introduce [semgroup](https://github.com/fatih/semgroup) library for rate-limited concurrency (while traversing metrics x 2 times)
* introduce [backoff](https://github.com/cenkalti/backoff) library for retrying `v1api.Query()`
* use `sync.Map` and `sync.RWMutex` (for concurrency safety)
* move duplicate logics to basic functions (clean code practices)

cc @anilmisirlioglu

### PR DESCRIPTION
**Pull request by anilmisirlioglu**
#### What this PR does

This PR adds resiliency and concurrency in analyze Prometheus command. Removes duplicate codes and optimizes code. It reduces the working time considerably. Notably performance increment as you can see the following:

image__51_

#### Which issue(s) this PR fixes or relates to

Fixes #3062

#### Checklist

- [x] Tests updated
- [ ] Documentation added
- [x] `CHANGELOG.md` updated - the order of entries should be `[CHANGE]`, `[FEATURE]`, `[ENHANCEMENT]`, `[BUGFIX]`

cc: @Dentrax @aysenrsrt

### CODE REVIEW

**Comment by osg-grafana** [comment]
Unchecked documentation in description because it does not yet exist; does this change need documentation? Will a user know to what extent it could be used?

**Comment by Dentrax** [comment]
Since we've respected as-is design during refactor, this PR does not include any behavioural change from the end user perspective. (Only `--concurrent` flag, which is just introduced). So no documentation needed IMO.

File: pkg/mimirtool/commands/analyse_prometheus.go [hunk 1]
```diff
@@ -34,173 +38,197 @@ type PrometheusAnalyzeCommand struct {
...
+	log.Infof("%d active series are NOT being used in dashboards", additionalMetricsAnalyse.cardinality)
+
+	metricsInPrometheus := cmd.result(inUseMetricsAnalyse, additionalMetricsAnalyse)
+	log.Infof("%d in use active series metric count", len(metricsInPrometheus.InUseMetricCounts))
+	log.Infof("%d additional active series metric count", len(metricsInPrometheus.AdditionalMetricCounts))
+
+	return cmd.write(metricsInPrometheus)
+}
+
+func (cmd *PrometheusAnalyzeCommand) parse() (model.LabelValues, error) {
```
Comment (Logiraptor) [inline pkg/mimirtool/commands/analyse_prometheus.go:88]: Can we use a more descriptive name here, maybe `parseUsedMetrics`?

File: pkg/mimirtool/commands/analyse_prometheus.go [hunk 2]
```diff
@@ -34,173 +38,197 @@ type PrometheusAnalyzeCommand struct {
...
-	if !hasGrafanaMetrics && !hasRulerMetrics {
-		return errors.New("No Grafana or Ruler metrics files")
+	if len(metricsUsed) == 0 {
+		return nil, errors.New("no Grafana or Ruler metrics files")
 	}
 
+	return metricsUsed, nil
+}
+
+func (cmd *PrometheusAnalyzeCommand) init() (v1.API, error) {
```
Comment (Logiraptor) [inline pkg/mimirtool/commands/analyse_prometheus.go:112]: Can we use more descriptive name here as well, maybe `initAPI` or `newAPI`?

File: pkg/mimirtool/commands/analyse_prometheus.go [hunk 3]
```diff
@@ -34,173 +38,197 @@ type PrometheusAnalyzeCommand struct {
...
+	}
+	log.Infof("%d active series are being used in dashboards", inUseMetricsAnalyse.cardinality)
+
+	log.Debugln("Starting to analyze additional metrics in use")
+	additionalMetricsAnalyse, err := cmd.analyze(apiWithBackoff, metricNames, func(metrics analyse) func(metric model.LabelValue) bool {
+		return func(metric model.LabelValue) bool {
+			_, ok := metrics.stats[metric]
+			return ok
+		}
+	}(inUseMetricsAnalyse))
```
Comment (Logiraptor) [inline pkg/mimirtool/commands/analyse_prometheus.go:75]:
```suggestion
	additionalMetricsAnalyse, err := cmd.analyze(apiWithBackoff, metricNames, func(metric model.LabelValue) bool {
			_, ok := inUseMetricsAnalyse.stats[metric]
			return ok
	})
```

This can be simplified to avoid the second closure, right?

File: pkg/mimirtool/commands/analyse_prometheus.go [hunk 4]
```diff
@@ -34,173 +38,197 @@ type PrometheusAnalyzeCommand struct {
...
-				counts.jobCount = make(map[string]int)
-			}
+	log.Infof("Found %d metric names\n", len(metricNames))
+	return metricNames, nil
+}
 
-			counts.totalCount += int(sample.Value)
-			counts.jobCount[string(sample.Metric["job"])] += int(sample.Value)
-			inUseMetrics[metric] = counts
+type analyse struct {
```
Comment (Logiraptor) [inline pkg/mimirtool/commands/analyse_prometheus.go:147]: Can we use `analysisResult` or another noun instead of `analyze`?

Comment (Logiraptor) [reply pkg/mimirtool/commands/analyse_prometheus.go:147]: If so, can you update the variables like `inUseMetricsAnalyze` to `inUseMetricsAnalysisResult` for example?

File: pkg/mimirtool/commands/analyse_prometheus.go [hunk 5]
```diff
@@ -34,173 +38,197 @@ type PrometheusAnalyzeCommand struct {
...
-		totalCount int
-		jobCount   map[string]int
-	}{}
-	additionalMetricsCardinality := 0
-	for _, metricName := range metricNames {
-		metric := string(metricName)
-		if _, ok := inUseMetrics[metric]; ok {
-			continue
+	// Concurrently fetch rules from all rulers. Since rules are not replicated,
+	// we need all requests to succeed.
```
Comment (Logiraptor) [inline pkg/mimirtool/commands/analyse_prometheus.go:169]: I think this is a copy paste right? It doesn't seem relevant to the code here.

**Comment by anilmisirlioglu** [comment]
Thanks! Just resolved your reviews, can you please check again?

File: pkg/mimirtool/commands/analyse_prometheus.go [hunk 6]
```diff
@@ -212,37 +254,42 @@ func (cmd *PrometheusAnalyzeCommand) run(k *kingpin.ParseContext) error {
...
-		sort.Slice(jobCounts, func(i, j int) bool {
-			return jobCounts[i].Count > jobCounts[j].Count
-		})
-		output.AdditionalMetricCounts = append(output.AdditionalMetricCounts, analyze.MetricCount{Metric: metric, Count: counts.totalCount, JobCounts: jobCounts})
+	return metricCount
+}
+
+func parseMetricFileIfExist(path string, out any) error {
+	if _, err := os.Stat(path); err != nil {
+		// if given file is not exist, it's OK to skip.
```
Comment (dimitarvdimitrov) [inline pkg/mimirtool/commands/analyse_prometheus.go:269]:
```suggestion
		// if given file does not exist, it's OK to skip.
```

**Review by dimitarvdimitrov — COMMENTED** [review]
Amazing, thanks for working on this! I have a few comments, most of which are minor

File: pkg/mimirtool/commands/analyse_prometheus.go [hunk 7]
```diff
@@ -34,173 +38,211 @@ type PrometheusAnalyzeCommand struct {
...
 	if err != nil {
-		return errors.Wrap(err, "error querying for metric names")
+		return nil, errors.Wrap(err, "error querying for metric names")
 	}
+
 	log.Infof("Found %d metric names\n", len(metricNames))
+	return metricNames, nil
+}
+
+func (cmd *PrometheusAnalyzeCommand) analyze(api v1.API, metrics model.LabelValues, filter func(model.LabelValue) bool) (analyseResult, error) {
```
[the remark this answers is not in the document]
Comment (anilmisirlioglu) [reply pkg/mimirtool/commands/analyse_prometheus.go]: Actually, we respect the previous condition, we can not check the length since the `stats[metric]` is a struct: https://github.com/grafana/mimir/blob/b9a7d2b87827e09c2d1f8ae6f42ba025e9287f46/pkg/mimirtool/commands/analyse_prometheus.go#L151-L154

File: pkg/mimirtool/commands/analyse_prometheus.go [hunk 8]
```diff
@@ -34,173 +38,211 @@ type PrometheusAnalyzeCommand struct {
...
+	}
+
+	log.Debugln("Starting to analyze metrics in use")
+	inUseMetricsAnalysisResult, err := cmd.analyze(v1api, metricsUsed, func(model.LabelValue) bool { return false })
+	if err != nil {
+		return err
+	}
+	log.Infof("%d active series are being used in dashboards", inUseMetricsAnalysisResult.cardinality)
+
+	log.Debugln("Starting to analyze additional metrics in use")
```
[the remark this answers is not in the document]
Comment (anilmisirlioglu) [reply pkg/mimirtool/commands/analyse_prometheus.go]: We renamed this but we didn't change the output struct to not introduce a breaking change.

[the remark this answers is not in the document]
Comment (dimitarvdimitrov) [reply pkg/mimirtool/commands/analyse_prometheus.go]: yeah i figured. Keeping the JSON as-is is fine by me

File: pkg/mimirtool/commands/analyse_prometheus.go (hunk 7 above)
[the remark this answers is not in the document]
Comment (dimitarvdimitrov) [reply pkg/mimirtool/commands/analyse_prometheus.go]: I was thinking about this

https://github.com/grafana/mimir/blob/b9a7d2b87827e09c2d1f8ae6f42ba025e9287f46/pkg/mimirtool/commands/analyse_prometheus.go#L169-L177

vs 

https://github.com/anilmisirlioglu/mimir/blob/38d47bc16bd41e9abcfa478938db2246d04fc070/pkg/mimirtool/commands/analyse_prometheus.go#L175-L179


---

you can do 
```go
		return inUseMetricsAnalysisResult.stats[metric].totalCount > 0
```

**Review by dimitarvdimitrov — APPROVED** [review]
apart from that minor godoc/usage difference, LGTM!

File: pkg/mimirtool/commands/analyse_prometheus.go (hunk 7 above)
[the remark this answers is not in the document]
Comment (dimitarvdimitrov) [reply pkg/mimirtool/commands/analyse_prometheus.go]: actually changing that return wouldn't retain the old behaviour. But I can't think of a reason why the query wouldn't return any results (i.e. empty vector), but we would still want to count the metric as used. Could the old behaviour be incorrect?

[the remark this answers is not in the document]
Comment (anilmisirlioglu) [reply pkg/mimirtool/commands/analyse_prometheus.go]: You are right, we changed this line

**Comment by anilmisirlioglu** [comment]
I run and commit it. I hope the linter is fixed.

File: CHANGELOG.md [hunk 9]
```diff
@@ -63,6 +63,8 @@
 ### Mimirtool
 
 * [ENHANCEMENT] Added `mimirtool rules delete-namespace` command to delete all of the rule groups in a namespace including the namespace itself. #3136
+* [ENHANCEMENT] Refactor `mimirtool analyze prometheus`: add concurrency and resiliency #3062
```
Comment (pracucci) [inline CHANGELOG.md:66]: We're used to reference the PR number instead of the issue number (so `#3349` instead of `#3062`). This will make our release tooling matching the PR when checking the CHANGELOG.

@anilmisirlioglu would you mind opening a follow up PR to change it, please?

Comment (anilmisirlioglu) [reply CHANGELOG.md:66]: Sure
